src/server/system_services/system_store.js

```diff
@@ -589,8 +589,11 @@ class SystemStore extends EventEmitter {
      *
      */
     async make_changes(changes) {
+        // Refreshing must be done outside the semapore lock because refresh
+        // might call load that is locking on the same semaphore.
+        const data = await this.refresh();
```

Comment:
I think we don't want to save the data from before the lock and send it inside, because we might wait for the lock a bit, and the data we took might be stale.

I think we can call `await refresh()` outside the lock, and inside _make_changes_internal we would just take the latest data from `const data = this.data`. 

If we want to be super duper, we can add a check inside the lock that if `Date.now()-this.data.time > this.FORCE_REFRESH_THRESHOLD` then we warn/throw? (also if > START_REFRESH_THRESHOLD we better warn...)

Reply:
also, we can add a warning timeout to this semaphore like you suggested in the issue, just in case.

Reply:
that's right, thanks!

Reply:
I had many directions about what is a correct warning timeout but I left with thinking that in the case of semaphore count is 1, work timeout should be equal waiting outside timeout, so if we start warn after START_REFRESH_THRESHOLD we assume that load() work takes less than START_REFRESH_THRESHOLD, not sure about it, let me know what you think